### PR TITLE
Suggest the minimal type disambiguation when an overload doesn't have any unique types

### PR DESCRIPTION
**Pull request by d-ronnqvist**
Bug/issue #, if applicable: rdar://136207880

## Summary

This improves the suggested parameter type and return value type disambiguation to suggest the minimal amount of disambiguation when an overload doesn't have any unique types.

For example, consider these 3 overloads:
```swift
func doSomething(first: String,  second: Int, third: Double) {}
func doSomething(first: String?, second: Int, third: Double) {}
func doSomething(first: String?, second: Int, third: Float)  {}
```

Currently, because the first overload is the only one with `String` as its first parameter type and the third overload is the one one with `Float` as its last parameter type, DocC suggests readable type signature disambiguation for those two overloads. However, because the second overload requires a combination of two type names to uniquely disambiguate it, DocC doesn't currently suggest a type signature for disambiguation and instead falls back to a hash for disambiguation:

Screenshot 2024-11-06 at 13 46 54

**Note:** DocC can unambiguously resolve ` ``doSomething(first:second:third:)-(String?,_,Double)`` ` as a link to the second overload but it doesn't make that suggestion when the link is ambiguous. 

With the enhancements in this PR, DocC will find the fewest and shortest type names that uniquely disambiguate each overload and suggest that:

Screenshot 2024-11-06 at 13 53 56

Also, with the enhancements in this PR, DocC is able to suggest type signature disambiguation that _combines_ parameter types and return types. 

For example, consider these 3 overloads (where the previous third parameter is a return value instead):
```swift 
func doSomething(first: String,  second: Int) -> Double { 0.0 }
func doSomething(first: String?, second: Int) -> Double { 0.0 }
func doSomething(first: String?, second: Int) -> Float  { 0.0 }
```

The first overload is the only with a `String` type for its first parameter and the third overload is the only with a `Float` return value but the second overload doesn't have either unique parameter types (same as the third overload) or return types (same as the first overload). In this case DocC suggests a combination of parameter types and return types (`-(String?,_)->Double`) to uniquely disambiguate the second overload:

Screenshot 2024-11-06 at 13 58 57

### Details

The [first commit of this PR](https://github.com/swiftlang/swift-docc/commit/6109d286aecc4b7d982977e771a4c8e2e4454077) implemented the general support for suggesting combinations of type names as disambiguation and [the second commit of this PR](https://github.com/swiftlang/swift-docc/commit/4d2d4828c2b7775e15ce061c6af52c43b48a0474) extended that support to combinations of parameter types and return value types. All the other changes in this PR is to ensure that this code is very fast.

#### Performance

To measure the performance of this code, I extracted an example overload group of 8 symbols with 6 parameters from a project and computed the suggested disambiguation for an ambiguous link to that symbol 50k times. This is a rather high number to measure since it's very unlikely that a project would have 50k ambiguous links. 

However, because `docc process-catalog emit-generated-curation` computes the suggested disambiguation for every symbol in a project, it's possible that this code would run a _few_ thousands of times. (Running 50k times would require that the target has 50k distinct _overload groups_ which would require at least 100k symbols (if each group is only two symbols) or many hundreds of thousands of symbols if each overload groups contains more symbols).

Initially, calling `PathHierarchy.DisambiguationContainer.disambiguatedValues(...)` for these 8 symbols 50k times took just short of 3 seconds. 

The initial bottleneck was my naive code for computing the sequence of type name combinations to try. [Using the optimized combinations code in Swift Algorithms](https://github.com/swiftlang/swift-docc/commit/7684ae0cd0562a710d77a8294983f58a41e69853) brought this down to about 2 seconds.

After that, most of the time was spent accessing and modifying `Set<Int>` to track which overloads' type names occur in other overloads and to track which type names to try to use for disambiguation. I realized that the numbers these sets were tracking were always in the range `0 ..< numberOfTypeNames`. 

My intuition is that most functions/initializers/subscripts/etc. only have a handful of parameters and that the number of symbols quickly goes down as the number of parameters does up. Tangential data seems to support this. For example, the number of symbols with different number of parameters (greater than 0) in the Swift standard library looks like this with a logarithmic Y-axis:

Screenshot 2024-11-06 at 15 13 15

(The 8, 16, 32, and 64-parameter outliers are each a single symbol (the element-wise initializers for [SIMD8](https://developer.apple.com/documentation/swift/simd8/init(_:_:_:_:_:_:_:_:)), [SIMD16](https://developer.apple.com/documentation/swift/simd16/init(_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:)), [SIMD32](https://developer.apple.com/documentation/swift/simd32/init(_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:)), and [SIMD64](https://developer.apple.com/documentation/swift/simd64/init(_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:_:))) meaning that there's no other symbols with the same number of parameters to disambiguate from, ignoring that the symbols would also need to have the same names to need disambiguation)

The same distribution—also with a logarithmic Y-axis—for Foundation looks like this:

Screenshot 2024-11-06 at 15 15 20

For an unnamed C library with lots of overloads it looks like this:

Screenshot 2024-11-06 at 15 17 13

And for two large unnamed frameworks the distributions look like this:

Screenshot 2024-11-06 at 15 18 31

Screenshot 2024-11-06 at 15 21 22

This anecdotal data seems to support the idea that an optimization specific to overloads with "few" parameters would be applicable in the common case.

Since the common optimized case only needs to consider the numbers `0 ..< 64`, I implemented [a specialized set-algebra type](https://github.com/swiftlang/swift-docc/commit/d9864caaf20527f8ad6153979f391ce29f525638) that stores each number in a `UInt64` bit set, which brought down the time to about 1.1 second.

Now, most of the time was spent computing the sequence of combinations to try. By leveraging the fact that all possible combinations of the specialized bit-set type can be generated by looping over a range of integers and sorting them by `.nonzeroBitCount`, a [specialized combinations generator](https://github.com/swiftlang/swift-docc/commit/b010bb9933fa9600a9b66786ef716948a0706a47) could bring down the time to about 850 ms.

The next noteworthy bottleneck was creating and accessing the nested `[[IntSet]]` to determine if a combination of type names is ambiguous or not. I added [a specialized `Table<Element>` type](https://github.com/swiftlang/swift-docc/commit/4ce6917e2381030ce6b15bc210c0665c214ad91e) that stores its elements in a single contiguous storage, which brought down the time to about 600 ms.

This was already 5 times faster than the original implementation, and I should have probably stopped here. However, through a series of many smaller improvements I brought the time down to about 350 ms. 

At this point, only 25% of the time spent is computing the minimal disambiguation and the remaining 75% is in higher level code that:
- tracks which elements have already computed disambiguation so that we don't suggest type disambiguation for an element that can be disambiguated by its symbol kind
- attempts to disambiguate by symbol kind
- groups elements by number of parameters so that the low level code can compute the minimal disambiguation for each group.

This is where I decided to stop 😄

Despite all the low level optimizations I still find the code _fairly_ easy to read and follow. We'll see how I feel about that in a year 🤞. 

## Dependencies

This adds Swift Algorithms as a package dependency. 

However, it is only used in the very rare case when a group of overloaded symbols have more than 64 parameters or when there are more than 64 different symbols in an overload group. If we prefer to not add a new dependency we could remove it and use a different implementation for large overload groups with large numbers of parameters. 

For example, we could consider only checking those 64+ parameter overloads for disambiguation involving either one or two different parameters. Such a combinations sequence is not too hard to implement and the only impact would be that the very very _very_ rare case would fallback to _suggesting_ hash disambiguation when we could have found a type disambiguation if we checked more combinations. 

We could also consider not even trying to find type disambiguation for 64+ parameters because arguably, `-(_,_,_,_,_,_,_,_,_,_,_,_,_,String,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,Int,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,_,)` has its own readability problems

## Testing

1. Define 3 functions (either top-level or within a class or struct) like these
  ```swift
  func doSomething(first: String,  second: Int, third: Double) {}
  func doSomething(first: String?, second: Int, third: Double) {}
  func doSomething(first: String?, second: Int, third: Float)  {}
  ```
2. From some other symbol, write a link to `doSomething(first:second:third:)`. 
  - The warning about the ambiguous link should suggest parameter disambiguation for all 3 overloads.
    Screenshot 2024-11-06 at 13 53 56

3. Define 3 new functions (either top-level or within a class or struct) like these
  ```swift
  func doSomething(first: String,  second: Int) -> Double { 0.0 }
  func doSomething(first: String?, second: Int) -> Double { 0.0 }
  func doSomething(first: String?, second: Int) -> Float  { 0.0 } 
  ```
4. From some other symbol, write a link to `doSomething(first:second:)`. 
  - The warning about the ambiguous link should suggest parameter disambiguation, return value disambiguation, or combined parameter and return value disambiguation for all 3 overloads.
    Screenshot 2024-11-06 at 13 58 57

## Checklist

Make sure you check off the following items. If they cannot be completed, provide a reason.

- [x] Added tests
- [x] Ran the `./bin/test` script and it succeeded
- ~[ ] Updated documentation if necessary.~ I'll add documentation separately. This is tracked by rdar://136207820

### CODE REVIEW

**Comment by QuietMisdreavus** [comment]
This is impressive! I'll try to thumb through this today to give it a review.

> This adds Swift Algorithms as a package dependency.

TBH i'm fine with sneaking this in, if only because it adds the opportunity to clean up a little bit of the code i added in #928 😅

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+            }
+            
+            guard let (indicesToInclude, _) = shortestDisambiguationSoFar else {
+                // This overload can't be uniquely disambiguated by these type names
+                return nil
+            }
+            
+            // Found the fewest (and shortest) type names that uniquely disambiguate this overload.
+            // Return the list of disambiguating type names or "_" for an unused type name.
+            return typeNames.columnIndices.map {
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift:179]: Is it fewest or shortest? Reading the code above (the `reduce(0)` snippet that calculates the length) it seems like this algorithm returns the overall shortest string, and not necessarily the string which contains the fewest number of disambiguating types. Is this correct?

Maybe refine the comment here to explain.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+    // Explicitly specify the associated types for Swift 5.9 compatibility.
+    typealias Combination = [Int]
+    typealias CombinationSequence = Algorithms.CombinationsSequence<Self>
+    
+    func combinationsToCheck() -> CombinationSequence {
+        // For `Set<Int>`, use the Swift Algorithms implementation to generate the possible combinations.
+        self.combinations(ofCount: 1...)
+    }
+}
+extension _TinySmallValueIntSet: _IntSet, _Combination {
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift:218]: Why do you declare the protocol extension first, before ever defining the `_TinySmallValueIntSet` struct? Maybe reorder the code here a bit to make it easier to follow?

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+
+        private func index(row: Int, column: Int) -> Int {
+            // Let the wrapped buffer validate the index
+            row * width + column
+        }
+    }
+
+    @inlinable
+    subscript(row: Int, column: Int) -> Element {
+        _read { yield storage[index(row: row, column: column)] }
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift:417]: What does `_read` do here? And what about `yield` ? I'm not familiar with these. Are you handling multithreaded access to `storage`?

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+    let size: Size
+    private let storage: ContiguousArray<Element>
+
+    @inlinable
+    init(width: Int, height: Int, initializingWith initializer: (_ buffer: inout UnsafeMutableTableBufferPointer) throws -> Void) rethrows {
+        size = (width, height)
+        let capacity = width * height
+        storage = try .init(unsafeUninitializedCapacity: capacity) { buffer, initializedCount in
+            var wrappedBuffer = UnsafeMutableTableBufferPointer(width: width, wrapping: buffer)
+            try initializer(&wrappedBuffer)
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift:390]: I'm surprised that Swift allows you to call a closure like this, while executing an initializer, passing in a pointer to the storage.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+        private var wrapping: UnsafeMutableBufferPointer<Element>
+
+        init(width: Int, wrapping: UnsafeMutableBufferPointer<Element>) {
+            self.width = width
+            self.wrapping = wrapping
+        }
+
+        @inlinable
+        func initializeElementAt(row: Int, column: Int, to element: Element) {
+            wrapping.initializeElement(at: index(row: row, column: column), to: element)
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift:406]: Clever trick to wrap the underlying storage and make it appear as it it were a 2-D array.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift [hunk 6]
```diff
@@ -0,0 +1,437 @@
...
+    }
+    
+    @inlinable
+    func isSuperset(of other: Self) -> Bool {
+        // Provide a custom implementation since this is called frequently in `combinationsToCheck()`
+        (storage & other.storage) == other.storage
+    }
+    
+    @inlinable
+    func union(_ other: Self) -> Self {
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift:313]: You need to implement all of these to conform to `SetAlgebra`? It's amazing how well these operations map to the bit mask operations.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+    mutating func remove(_ member: Int) -> Int? {
+        let newStorage = storage & ~Self.mask(member)
+        defer {
+            storage = newStorage
+        }
+        return newStorage != storage ? member : nil
+    }
+    
+    @inlinable
+    @discardableResult
```
Comment (patshaughnessy) [inline Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift:290]: Why do we need this behavior for returning the removed value? Just to conform to `SetAlgebra`? I don't see any explicit dependency on this here in this PR...?

**Review by patshaughnessy — APPROVED** [review]
Amazing work!

Thank you for the series of separate commits explaining your progress improving the performance. We could convert this into a sample app on Swift performance - you used so many different techniques here!

I left a few questions and my usual readability suggestions. The only real suggestion I have, which you alluded to in the PR description, is that maybe it's not worth adding Swift Algorithms just for a few extremely unusual scenarios (symbols with over 64 parameters or return types) that we won't hit very often. When we do hit them, maybe an old-fashioned disambiguation hash would be fine. Aside from avoiding the new dependency, if you remove that bit you could also avoid the complexity around the `_IntSet` and `_Combination` private protocols.

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+    /// can be disambiguated using:
+    ///  - `String,_,_` because only the first overload has `String` as its first type
+    ///  - `String?,_,Double` because the combination of `String?` as its first type and `Double` as the last type is unique to the second overload.
+    ///  - `_,_,Float` because only the last overload has `Float` as its last type.
+    ///
+    /// - Parameter listOfTypeNames: The lists of type-name lists to shrink to the minimal unique combinations of type names
+    static func minimalSuggestedDisambiguation(listOfOverloadTypeNames: [[String]]) -> [[String]?] {
+        // The number of types in each list
+        guard let numberOfTypes = listOfOverloadTypeNames.first?.count, 0 < numberOfTypes else {
+            return []
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift:36]: Sure, we could move the responsibility to the caller to not call this function if the overloads have no return values. The only current caller already does this check.

**Comment by d-ronnqvist** [comment]
@swift-ci please test

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+        guard !typeNameIndicesToCheck.isEmpty else {
+            // Every type name is common across all overloads. These type names can't be used to disambiguate the overloads.
+            return .init(repeating: nil, count: typeNames.size.width)
+        }
+        
+        // Create a sequence of type name combinations with increasing number of type names in each combination.
+        let typeNameCombinationsToCheck = typeNameIndicesToCheck.combinationsToCheck()
+        
+        return typeNames.rowIndices.map { row in
+            var shortestDisambiguationSoFar: (indicesToInclude: IntSet, length: Int)?
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift]: I explicitly initialized this variable in https://github.com/swiftlang/swift-docc/pull/1087/commits/562d502bd6613d06e10297765efd24716349af57

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+        // Check if any columns are common for all overloads. Those type names won't meaningfully disambiguate any overload.
+        let allOverloads = IntSet(typeNames.rowIndices)
+        let typeNameIndicesToCheck = IntSet(typeNames.columnIndices.filter {
+            // It's sufficient to check the first row because this column has to be the same for all rows
+            table[0, $0] != allOverloads
+        })
+        
+        guard !typeNameIndicesToCheck.isEmpty else {
+            // Every type name is common across all overloads. These type names can't be used to disambiguate the overloads.
+            return .init(repeating: nil, count: typeNames.size.width)
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift:138]: I rephrased this to mention that nil means that the overload can't be disambiguated using these type names in https://github.com/swiftlang/swift-docc/pull/1087/commits/562d502bd6613d06e10297765efd24716349af57

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift [hunk 11]
```diff
@@ -361,6 +383,9 @@ extension PathHierarchy.DisambiguationContainer {
             case .parameterTypes(let types):
                 // For example: "-(String,_)" or "-(_,Int)"` (a certain parameter has a certain type), or "-()" (has no parameters).
                 return "-(\(types.joined(separator: ",")))"
+                
+            case .mixedTypes(parameterTypes: let parameterTypes, returnTypes: let returnTypes):
+                return "\(Self.parameterTypes(parameterTypes).makeSuffix())\(Self.returnTypes(returnTypes).makeSuffix())"
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift]: I simplified this with `+` in https://github.com/swiftlang/swift-docc/pull/1087/commits/562d502bd6613d06e10297765efd24716349af57

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift (hunk 8 above)
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+TypeSignatureDisambiguation.swift:39]: I added an assertion here in I explicitly labeled these closure arguments in https://github.com/swiftlang/swift-docc/pull/1087/commits/562d502bd6613d06e10297765efd24716349af57

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift [hunk 12]
```diff
@@ -233,7 +240,7 @@ extension PathHierarchy.DisambiguationContainer {
             collisions += _disambiguateByTypeSignature(
                 elementsThatSupportAdvancedDisambiguation,
                 types: \.returnTypes,
-                makeDisambiguation: Disambiguation.returnTypes,
+                makeDisambiguation: { .returnTypes($1) },
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift]: I explicitly labeled these closure arguments in https://github.com/swiftlang/swift-docc/pull/1087/commits/562d502bd6613d06e10297765efd24716349af57

File: Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift [hunk 13]
```diff
@@ -382,36 +407,46 @@ extension PathHierarchy.DisambiguationContainer {
...
-                    var disambiguation = [String](repeating: "_", count: typesCount)
-                    disambiguation[typeIndex] = returnType
-                    collisions.append((value: element.node, disambiguation: makeDisambiguation(disambiguation)))
-                    remainingIDs.remove(element.node.identifier)
-                    continue
+            guard numberOfTypeNames > 0 else {
+                continue // Need at least one type name to disambiguate (when there are multiple elements without parameters or return values)
+            }
+            
+            let disambiguation = minimalSuggestedDisambiguation(forOverloadsAndTypeNames: elementAndTypeNamePairs)
```
[the remark this answers is not in the document]
Comment (d-ronnqvist) [reply Sources/SwiftDocC/Infrastructure/Link Resolution/PathHierarchy+DisambiguatedPaths.swift]: I changed to a plural name for this variable in I explicitly labeled these closure arguments in https://github.com/swiftlang/swift-docc/pull/1087/commits/562d502bd6613d06e10297765efd24716349af57

**Comment by patshaughnessy** [comment]
Great work 👏  - thanks for the tweaks, comments and explanations here in the PR! I'm going to bookmark this one :)

**Comment by d-ronnqvist** [comment]
@swift-ci please test

**Review by franklinsch — APPROVED** [review]

**Pull request merged**